src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -15799,34 +15801,208 @@ def stack(
+        index_names : list[Hashable]
+            The list of index names.
+        operation : StackOperation.STACK or StackOperation.UNSTACK
+            The operation being performed.
+        """
+        # Resetting the index keeps the index columns as the first n data columns
+        qc = self.reset_index()
+        index_cols = qc._modin_frame.data_column_pandas_labels[0 : len(index_names)]
+        # We need to track where the index and columns originally had no name in order to reset
+        # the those names back to None after the operation
+        unnamed_index_positions = [i for i, x in enumerate(index_names) if x is None]
+        column_names_to_reset_to_none = []
+        # Track the new column names for the original unnamed index
+        for i in unnamed_index_positions:
+            column_names_to_reset_to_none.append(
+                qc._modin_frame.data_column_pandas_labels[i]
+            )
+        # Track the new column name for the original unnamed column
+        if self.columns.name is None:
+            quoted_col_label = (
+                qc._modin_frame.ordered_dataframe.generate_snowflake_quoted_identifiers(
+                    pandas_labels=["index_second_level"]
+                )[0]
+            )
+            col_label = unquote_name_if_quoted(quoted_col_label)
+            column_names_to_reset_to_none.append(col_label)
+        else:
+            col_label = self.columns.name
+
+        qc = qc.melt(
+            id_vars=index_cols,  # type: ignore
+            value_vars=self.columns,
+            var_name=col_label,
+            value_name=MODIN_UNNAMED_SERIES_LABEL,
+            ignore_index=False,
+        )
+
+        if operation == StackOperation.STACK:
+            # Only sort rows by column values in case of 'stack'
+            qc = qc.sort_rows_by_column_values(
+                columns=index_cols,  # type: ignore
                 ascending=[True],
                 kind="stable",
                 na_position="last",
                 ignore_index=False,
             )
-            .replace(to_replace=UNPIVOT_NULL_REPLACE_VALUE, value=np.nan)
-            .set_index_from_columns(index_names + ["index_second_level"])  # type: ignore
-            .set_index_names([None, None])
-        )
 
-        if dropna:
-            return qc.dropna(axis=0, how="any", thresh=None)
+        # TODO: SNOW-1524695: Remove the following replace once "NULL_REPLACE" values are fixed for 'melt'
+        qc = qc.replace(to_replace=UNPIVOT_NULL_REPLACE_VALUE, value=np.nan)
+
+        if operation == StackOperation.STACK:
+            qc = qc.set_index_from_columns(index_cols + [col_label])  # type: ignore
         else:
-            return qc
+            qc = qc.set_index_from_columns([col_label] + index_cols)  # type: ignore
+
+        # Set the original unnamed index and column values back to None
+        output_index_names = qc.get_index_names()
+        output_index_names = [
+            None
```

Comment:
i felt you don't even need to record none here, it seems that you just want to reset the columns back to its orignal name, in other words,
orignial_col_label = None if self.columns.name is None else self.columns.name
so if operation == StackOperation.STACK:
        index_names = original_index_cols + [original_col_label]
    else:
       index_names = [origianl_col_label] + original_index_columns

is that what we want?

Reply:
It's not just the column name, we still need to track the index names that correspond to None and reset those as well.